### PR TITLE
Use text properties instead of overlays for ANSI coloring

### LINKED ISSUE #1452: ANSI coloring in REPL buffer is broken
cider

### How to reproduce

1. Interactively eval the following code:

```clj
  (println "foo")
  (println "\u001b[31mbar\u001b[0m")
  (println "baz")
```

2. Then you will see the following output in REPL buffer:

```
foo  <-- default face
bar  <-- red face
baz  <-- red face(!)
user>
```

### Cause

This is because new output of interactive-eval extends an overlay for ANSI coloring created before.

The following is a part of output of <kbd>C-u C-x =</kbd> at 'r' of 'bar' after evaluating `(println "\u001b[31mbar\u001b[0m")`:

```
There is an overlay here:
 From 1 to 4
  face                 (foreground-color . "#fb4934")
  modification-hooks   (ansi-color-freeze-overlay)
```

Then, the following is new one after evaluating `(println "baz")`:

```
There is an overlay here:
 From 1 to 8
  face                 (foreground-color . "#fb4934")
  modification-hooks   (ansi-color-freeze-overlay)
```

It cannot be probably said that it is due to CIDER because I suspect this is ansi-color's problem.
I think `ansi-color-freeze-overlay` [should prevent overlays from being extended](https://github.com/emacs-mirror/emacs/blob/acfb5cd0353406784f085ddb6edfb0d0587048c8/lisp/ansi-color.el#L488-L514) at the first place.
I've actually confirmed it works by modifying ansi-color to use `insert-behind-hooks` instead of `modification-hooks`.

However, I think it is not so difficult to introduce a workaround for this on CIDER's side.
For example, it seems to work for me to use `insert` instead of `insert-before-markers` [here](https://github.com/clojure-emacs/cider/blob/5cd56ce76bdc4b0c50962ce626521133fb9e6edb/cider-repl.el#L471) though I'm not sure whether this cause other problems.
(BTW, cider-repl-output-start/cider-repl-output-end seems not to work but it is another issue.)

I'm a novice at elisp (and CIDER internals, of course), so I would be happy if someone investigate this problem further.

### PR DESCRIPTION
**Pull request by rfkm**
### TL;DR

This PR improves the performance of REPL buffers by using text properties instead of overlays for ANSI coloring.
Also, this fixes #1452 while removing the hacky workaround introduced by #1455.

---

I recently sent some PRs related to overlays in the REPL buffer,
but I've finally realized we should avoid using overlays for ANSI coloring in the first place.
It's simply because overlays are pretty slow. I improved the performance a bit in #1488, but it's not radical.

So, in this PR, I remove all code related to overlays and use text properties instead.

I measured the performance by calling `cider-repl--emit-interactive-output` repeatedly.
The complete benchmark code is [here](https://gist.github.com/rfkm/c28f0e35965bcebe3e37), and the version of my Emacs is `GNU Emacs 24.4.1 (x86_64-apple-darwin14.0.0, NS apple-appkit-1343.14)`.

active

As this figure shows, the overlay version (current master) is getting slower and slower as the number of lines in the REPL buffer increases.

total

Total time the overlay version spent on inserting 1000 lines is about 5 seconds. As for the text property version, it's about 0.6 seconds.

If no window displays the REPL buffer, it is much faster in the both cases:

hidden

total_hidden

---

I keep my changes in #1488 as is because they are harmless, but I can remove them if it is necessary.

I also wonder whether I should remove my old changelog entries related to overlays.
Since this PR removes old PRs' changes, it looks a bit strange that these changelog entries appears in the same release.

P.S.

While I had not intended that, the PR #1455 seems to have improved the performance significantly ;)

active_old
(This is the worst case, that is, all outputs have a trailing overlay.)

### CODE REVIEW

**Comment by Malabarba** [comment]
Makes sense to me. Overlays are indeed very slow in large numbers. 
Thanks for the very thorough analysis. :-)

Do we have tests for this ansi coloring feature?

**Comment by bbatsov** [comment]
> I also wonder whether I should remove my old changelog entries related to overlays.
> Since this PR removes old PRs' changes, it looks a bit strange that these changelog entries appears in the same release.

Yeah, you should remove them.

Overall - impressive work. I was amazed by your comprehensive analysis of the problem!
Performance optimization work is critical, but we've always ignored it to some extent so far. Maybe you can help us with tickets like:

* https://github.com/clojure-emacs/cider/issues/228
* https://github.com/clojure-emacs/cider/issues/1115

> Do we have tests for this ansi coloring feature?

Pretty sure we don't. A few tests would be nice to have.

**Comment by Malabarba** [comment]
Also, I forgot to mention, but the code is good to merge for me (though some tests would be nice).

**Comment by bbatsov** [comment]
Yeah, it's good to merge for sure - I just want us to update the changelog accordingly as well.

**Comment by rfkm** [comment]
Sorry for the delay.
I've just pushed some changes.
Since this PR also fixes #1452, I kept the changelog entry about #1452 though I removed the mention of overlays.
Instead, I've added summary of this PR on top of this thread.
Also, I added a test for this.

> Performance optimization work is critical, but we've always ignored it to some extent so far. Maybe you can help us with tickets like:

I've already noticed `cider-repl--show-maximum-output` could be slow under certain conditions, but I hope one of the causes has been already eliminated.
One of the causes I noticed is `(recenter -1)` that `cider-repl--show-maximum-output` calls could be pretty slow if there are many overlays.
Also, I've noticed eldoc could be very slow in a large REPL buffer. As for this, I might send a PR in the near future.
Anyway, I'll try to ivenstigate their issues :)

**Comment by bbatsov** [comment]
Great! Thanks a lot! :-)

**Comment by Malabarba** [comment]
Thanks indeed!
